services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/geometria/HankeGeometriatDaoImpl.kt

```diff
@@ -0,0 +1,155 @@
+            val id = queryForObject("""
+                INSERT INTO HankeGeometriat (
+                    hankeId,
+                    version,
+                    createdByUserId,
+                    createdAt,
+                    modifiedByUserId,
+                    modifiedAt
+                ) VALUES (
+                    ?,
+                    ?,
+                    ?,
+                    ?,
+                    ?,
+                    ?               
+                )
+                RETURNING id
+                """.trimIndent(), { rs, _ -> rs.getInt(1) },
+                    hankeGeometriat.hankeId,
+                    hankeGeometriat.version ?: 0,
+                    hankeGeometriat.createdByUserId,
+                    if (hankeGeometriat.createdAt != null) Timestamp(hankeGeometriat.createdAt!!.toInstant().toEpochMilli()) else null,
+                    hankeGeometriat.modifiedByUserId,
+                    if (hankeGeometriat.modifiedAt != null) Timestamp(hankeGeometriat.modifiedAt!!.toInstant().toEpochMilli()) else null
+            )
+            hankeGeometriat.id = id
+            saveHankeGeometriaRows(hankeGeometriat, this)
+        }
+    }
+
+    override fun updateHankeGeometriat(hankeGeometriat: HankeGeometriat) {
+        with(jdbcOperations) {
+            update("""
+                UPDATE HankeGeometriat
+                SET
+                    version = ?,
+                    modifiedByUserId = ?,
+                    modifiedAt = ?
+                WHERE
+                    id = ?
+            """.trimIndent()) { ps ->
+                ps.setInt(1, hankeGeometriat.version!!)
+                if (hankeGeometriat.modifiedByUserId != null) {
+                    ps.setInt(2, hankeGeometriat.modifiedByUserId!!)
+                } else {
+                    ps.setNull(2, Types.INTEGER)
+                }
+                if (hankeGeometriat.modifiedAt != null) {
+                    ps.setTimestamp(3, Timestamp(hankeGeometriat.modifiedAt!!.toInstant().toEpochMilli()))
+                } else {
+                    ps.setNull(3, Types.TIMESTAMP)
+                }
+                ps.setInt(4, hankeGeometriat.id!!)
+            }
+            // delete old geometry rows
+            deleteHankeGeometriaRows(hankeGeometriat, this)
+            // save new geometry rows
+            saveHankeGeometriaRows(hankeGeometriat, this)
+        }
+    }
+
+    override fun retrieveHankeGeometriat(hankeId: Int): HankeGeometriat? {
+        with(jdbcOperations) {
+            val hankeGeometriat = query("""
+            SELECT
+                id,
+                hankeId,
+                version,
+                createdByUserId,
+                createdAt,
+                modifiedByUserId,
+                modifiedAt
+            FROM HankeGeometriat WHERE hankeId = ?            
+        """.trimIndent(), { rs, _ ->
+                HankeGeometriat(
+                        rs.getInt(1),
```

Comment:
Should we use local constants for readability... numbered "hihavakio"s are always a bit ugly.

Reply:
Those .getInt(1) etc, are pretty standard way to access the results.  The alternate version gets quite verbose.

Reply:
what do you mean by "constants"? rs.getInt(1)? They are not constants, they are indexes of ResultSet values.